### PR TITLE
Reporting and banning.

### PR DESCRIPTION
**Pull request by dkorolev**

### CODE REVIEW

File: schema.h [hunk 1]
```diff
@@ -437,6 +437,86 @@ struct CommentFlagAsInappropriate : Super {
...
+
+  template <typename A>
+  void serialize(A& ar) {
+    Super::serialize(ar);
+    ar(CEREAL_NVP(who), CEREAL_NVP(whom));
+  }
+};
+
+// To enable banned users support.
+struct BannedUser : Super {
```
Comment (mzhurovich) [inline schema.h:505]: Just curious, why is it better than adding a flag to `User` struct?

File: server.h [hunk 2]
```diff
@@ -955,17 +978,23 @@ class CTFOServer final {
...
               comments_mutator.Add(comment);
 
-              // Emit the "new comment" notification.
+              // Emit the "new comment on my card" notification.
               if (card_author_uid != UID::INVALID_USER && card_author_uid != uid) {
-                data.notifications.Add(Notification(
-                    card_author_uid, now, std::make_shared<NotificationMyCardNewComment>(comment)));
+                // Do not send "new comment on my card" if it is the reply to my own comment.
+                if (parent_comment_author_uid == UID::INVALID_USER ||
+                    parent_comment_author_uid != card_author_uid) {
```
Comment (mzhurovich) [inline server.h:984]: Just the second condition only should work, no?

File: schema.h (hunk 1 above)
Comment (mzhurovich) [inline schema.h:505]: Disregard, pls. The list of banned users should be easily accessible :)

File: server.h [hunk 3]
```diff
@@ -1107,21 +1139,36 @@ class CTFOServer final {
...
+          if (Exists(iterable)) {
+            const auto v = Value(iterable);
+            if (v.Size() == 1u) {
+              return (*v.begin()).uid;
+            }
+          }
+          return UID::INVALID_USER;
+        }(card.cid);
+
+        if (card_author_uid == UID::INVALID_USER || (!Exists(data.banned_users[card_author_uid]) &&
```
Comment (mzhurovich) [inline server.h:1153]: `card_author_uid == UID::INVALID_USER` - could it ever happen?

File: server.h (hunk 2 above)
Comment (dkorolev) [inline server.h:984]: Yes. Just like clarity sometimes. What if whoever reads this code is a Java guy who forgot how to use their brains years ago?

File: server.h (hunk 3 above)
Comment (dkorolev) [inline server.h:1153]: It can not happen now, just to stay on the safe side.

File: schema.h (hunk 1 above)
Comment (dkorolev) [inline schema.h:505]: :-)

**Pull request merged**